File: src/core/sync.ts
import { mkdir, readdir, readFile } from 'node:fs/promises';
import * as posix from 'node:path/posix';
import type { WorkshopId } from './acf.ts';
import type { Config } from './config.ts';
import { AxeError } from './errors.ts';
import { atomicWrite } from './io.ts';
import { type Layout, WORKSHOP_APPID } from './layout.ts';
import { runModsCheck } from './mods.ts';
import { runSteamcmd, type SpawnLike, type SteamcmdAction } from './steamcmd.ts';
import type { FetchLike } from './workshop.ts';

export type SyncOutcome = {
  /** declared IDs that produced .pak files in the workshop content tree */
  downloaded: WorkshopId[];
  /** declared IDs that produced no .pak files (still drift after sync) */
  missing: WorkshopId[];
  modlist_path: string;
  /** false when modlist bytes already matched the expected output */
  modlist_changed: boolean;
  log_file?: string;
  warnings: string[];
};

export type RunSyncOptions = {
  fetchImpl?: FetchLike;
  spawn?: SpawnLike;
  signal?: AbortSignal;
  /** path for steamcmd log capture; omitted = no log written */
  logFile?: string;
  /** override `[steamcmd].binary` config + PATH discovery */
  steamcmdBinary?: string;
};

/**
 * Sync workshop mods declared in `axe.toml`.
 *
 * Pipeline:
 *   1. `runModsCheck` to classify each declared id (current/stale/missing_local).
 *   2. If nothing is stale or missing_local, skip steamcmd entirely.
 *   3. Otherwise run a single batched `+workshop_download_item` steamcmd.
 *   4. Scan `<root>/steamapps/workshop/content/<appid>/<id>/` for `.pak` files.
 *   5. Build modlist text in declared order; byte-compare existing modlist;
 *      `atomicWrite` only if different.
 *
 * Throws `AxeError(config)` when steamcmd binary cannot be found and
 * downloads are required. Tolerates non-zero steamcmd exit (warning).
 */
export async function syncModlist(
  config: Config,
  layout: Layout,
  options: RunSyncOptions = {},
): Promise<SyncOutcome> {
  const declared = config.mods.ids;
  const warnings: string[] = [];

  const checkOptions: { fetchImpl?: FetchLike; signal?: AbortSignal } = {};
  if (options.fetchImpl !== undefined) checkOptions.fetchImpl = options.fetchImpl;
  if (options.signal !== undefined) checkOptions.signal = options.signal;
  const { report, warnings: checkWarnings } = await runModsCheck(layout, declared, checkOptions);
  warnings.push(...checkWarnings);

  const toDownload: WorkshopId[] = [];
  for (const item of report.items) {
    if (item.state === 'stale' || item.state === 'missing_local') toDownload.push(item.id);
  }

  // Fast path: freshness says nothing to do. Still verify the modlist on disk.
  if (toDownload.length === 0) {
    const modlist = await reconcileModlist(layout, declared);
    return {
      downloaded: [],
      missing: modlist.missing,
      modlist_path: layout.modlist_txt,
      modlist_changed: modlist.changed,
      warnings,
    };
  }

  const steamcmdBinary =
    options.steamcmdBinary ?? config.steamcmd?.binary ?? Bun.which('steamcmd') ?? null;
  if (!steamcmdBinary) {
    throw new AxeError(
      'config',
      'steamcmd binary not found; set [steamcmd].binary in axe.toml or install steamcmd on PATH',
    );
  }

  const actions: SteamcmdAction[] = toDownload.map((id) => ({
    kind: 'workshop_download_item',
    appid: WORKSHOP_APPID,
    id,
  }));

  const runOptions: {
    spawn?: SpawnLike;
    signal?: AbortSignal;
    logFile?: string;
  } = {};
  if (options.spawn !== undefined) runOptions.spawn = options.spawn;
  if (options.signal !== undefined) runOptions.signal = options.signal;
  if (options.logFile !== undefined) runOptions.logFile = options.logFile;

  const outcome = await runSteamcmd(
    {
      binary: steamcmdBinary,
      forceInstallDir: config.server.root,
      login: 'anonymous',
      actions,
    },
    runOptions,
  );

  if (outcome.exit !== 0) {
    const tail = outcome.stderr.trim().split('\n').slice(-3).join(' | ');
    warnings.push(
      `steamcmd exited ${outcome.exit}${tail ? `; tail: ${tail}` : ''}${
        options.logFile ? `; log: ${options.logFile}` : ''
      }`,
    );
  }

  const modlist = await reconcileModlist(layout, declared);

  // `downloaded`: ids we asked steamcmd for AND that produced .pak files.
  // `missing`: declared ids with no .pak (canonical post-sync drift set).
  const stillMissing = new Set(modlist.missing.map((id) => id.toString()));
  const downloaded = toDownload.filter((id) => !stillMissing.has(id.toString()));

  const result: SyncOutcome = {
    downloaded,
    missing: modlist.missing,
    modlist_path: layout.modlist_txt,
    modlist_changed: modlist.changed,
    warnings,
  };
  if (options.logFile !== undefined) result.log_file = options.logFile;
  return result;
}

/**
 * Build the modlist text for `declared` and write it iff bytes differ.
 * Returns `{ changed, missing }` where `missing` is declared ids whose
 * content directory has no `.pak` files (or doesn't exist).
 */
async function reconcileModlist(
  layout: Layout,
  declared: readonly WorkshopId[],
): Promise<{ changed: boolean; missing: WorkshopId[] }> {
  const missing: WorkshopId[] = [];
  const lines: string[] = [];
  for (const id of declared) {
    const paks = await scanPaks(layout.workshop_content, id);
    if (paks.length === 0) missing.push(id);
    for (const pak of paks) lines.push(pak);
  }
  const expected = lines.length > 0 ? `${lines.join('\n')}\n` : '';

  let current: string | null = null;
  try {
    current = await readFile(layout.modlist_txt, 'utf8');
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code !== 'ENOENT') {
      throw new AxeError('filesystem', `reading ${layout.modlist_txt}: ${(e as Error).message}`);
    }
  }

  if (current === expected) return { changed: false, missing };
  // skip writing an empty modlist if there's no file to begin with — avoid
  // creating noise on a fresh install with zero declared mods.
  if (expected === '' && current === null) return { changed: false, missing };

  await mkdir(layout.mods_dir, { recursive: true });
  await atomicWrite(layout.modlist_txt, expected);
  return { changed: true, missing };
}

/** Sorted list of `.pak` paths for a workshop id, empty if dir is missing. */
async function scanPaks(workshopContent: string, id: WorkshopId): Promise<string[]> {
  const dir = posix.join(workshopContent, id.toString());
  let entries: string[];
  try {
    entries = await readdir(dir);
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === 'ENOENT' || code === 'ENOTDIR') return [];
    throw new AxeError('filesystem', `reading ${dir}: ${(e as Error).message}`);
  }
  return entries
    .filter((name) => name.toLowerCase().endsWith('.pak'))
    .sort()
    .map((name) => posix.join(dir, name));
}

File: src/axe/sync.py
"""Drift reconciliation: mods + base build in one SteamCMD batch + hook lifecycle."""

from __future__ import annotations

import sys
from collections.abc import Sequence
from dataclasses import dataclass, field
from pathlib import Path

from axe.build import read_build_status
from axe.context import Context
from axe.errors import AxeError
from axe.hooks import HookRunner, SyncEnvContext, build_hook_env, run_hook
from axe.io import atomic_write
from axe.layout import SERVER_APPID, WORKSHOP_APPID, Layout
from axe.mods import run_mods_check
from axe.steamcmd import (
    AppUpdate,
    SpawnLike,
    SteamcmdAction,
    SteamcmdRequest,
    WorkshopDownloadItem,
    reserve_steamcmd_log,
    resolve_steamcmd,
    run_steamcmd,
)
from axe.workshop import FetchLike


@dataclass(frozen=True)
class SyncOutcome:
    downloaded: list[int]
    missing: list[int]
    modlist_path: str
    modlist_changed: bool
    log_file: Path | None
    base_build_updated: bool = False
    warnings: list[str] = field(default_factory=list)


def run_sync(
    ctx: Context,
    *,
    fetch: FetchLike | None = None,
    spawn: SpawnLike | None = None,
    hook_runner: HookRunner | None = None,
    log_file: Path | None = None,
    steamcmd_binary: str | None = None,
) -> SyncOutcome:
    """Reconcile any drift: mods + base build. Single steamcmd invocation when work needed."""
    declared = list(ctx.config.mods.ids)
    warnings: list[str] = []

    binary = steamcmd_binary or resolve_steamcmd(ctx.config.steamcmd.binary)

    mods_result = run_mods_check(ctx, fetch=fetch)
    warnings.extend(mods_result.warnings)
    build = read_build_status(ctx, spawn=spawn)
    warnings.extend(build.warnings)

    sync_env = SyncEnvContext(
        mods_stale=mods_result.report.stale,
        mods_missing_local=mods_result.report.missing_local,
        mods_missing_remote=mods_result.report.missing_remote,
        build=build,
    )
    # before_sync hook (strict-capable; raises AxeError('hook') on strict failure)
    pre_warn = run_hook(
        ctx.config.hooks.before_sync,
        build_hook_env(ctx, "before_sync", sync=sync_env),
        runner=hook_runner,
        strict=ctx.config.hooks.strict.before_sync,
    )
    if pre_warn:
        warnings.append(pre_warn)

    to_download = [
        item.id
        for item in mods_result.report.items
        if item.state in ("stale", "missing_local")
    ]

    actions: list[SteamcmdAction] = []
    if build.drifted:
        actions.append(AppUpdate(appid=SERVER_APPID, validate=False))
    actions.extend(WorkshopDownloadItem(appid=WORKSHOP_APPID, workshop_id=w) for w in to_download)

    if not actions:
        ml = _reconcile_modlist(ctx.layout, declared)
        outcome = SyncOutcome(
            downloaded=[],
            missing=ml.missing,
            modlist_path=str(ctx.layout.modlist_txt),
            modlist_changed=ml.changed,
            log_file=None,
            base_build_updated=False,
            warnings=warnings,
        )
        _fire_after_sync(ctx, sync_env, outcome, hook_runner, warnings)
        return outcome

    out_log = log_file or reserve_steamcmd_log(ctx.layout.root, "sync")
    if sys.stderr.isatty():
        print(f"running steamcmd sync (log: {out_log})", file=sys.stderr, flush=True)
    outcome_steamcmd = run_steamcmd(
        SteamcmdRequest(
            binary=binary,
            force_install_dir=str(ctx.layout.root),
            actions=actions,
        ),
        spawn=spawn,
        log_file=out_log,
        stream=True,
    )
    if outcome_steamcmd.exit != 0:
        tail = " | ".join(outcome_steamcmd.stderr.strip().splitlines()[-3:])
        warnings.append(
            f"steamcmd exited {outcome_steamcmd.exit}"
            + (f"; tail: {tail}" if tail else "")
            + f"; log: {out_log}"
        )

    ml = _reconcile_modlist(ctx.layout, declared)
    still_missing = set(ml.missing)
    downloaded = [w for w in to_download if w not in still_missing]

    outcome = SyncOutcome(
        downloaded=downloaded,
        missing=ml.missing,
        modlist_path=str(ctx.layout.modlist_txt),
        modlist_changed=ml.changed,
        log_file=out_log,
        base_build_updated=build.drifted and outcome_steamcmd.exit == 0,
        warnings=warnings,
    )
    _fire_after_sync(ctx, sync_env, outcome, hook_runner, warnings)
    return outcome


def _fire_after_sync(
    ctx: Context,
    sync_env: SyncEnvContext,
    outcome: SyncOutcome,
    runner: HookRunner | None,
    warnings: list[str],
) -> None:
    after_env = SyncEnvContext(
        mods_stale=sync_env.mods_stale,
        mods_missing_local=sync_env.mods_missing_local,
        mods_missing_remote=sync_env.mods_missing_remote,
        build=sync_env.build,
        outcome=outcome,
    )
    warn = run_hook(
        ctx.config.hooks.after_sync,
        build_hook_env(ctx, "after_sync", sync=after_env),
        runner=runner,
        strict=False,
    )
    if warn:
        warnings.append(warn)


@dataclass(frozen=True)
class _ModlistResult:
    changed: bool
    missing: list[int]


def _reconcile_modlist(layout: Layout, declared: Sequence[int]) -> _ModlistResult:
    missing: list[int] = []
    lines: list[str] = []
    for wsid in declared:
        paks = _scan_paks(layout.workshop_content, wsid)
        if not paks:
            missing.append(wsid)
        lines.extend(str(p) for p in paks)

    expected = ("\n".join(lines) + "\n") if lines else ""
    current: str | None
    try:
        current = layout.modlist_txt.read_text()
    except FileNotFoundError:
        current = None
    except OSError as e:
        raise AxeError("filesystem", f"reading {layout.modlist_txt}: {e}") from e

    if current == expected:
        return _ModlistResult(changed=False, missing=missing)
    if expected == "" and current is None:
        return _ModlistResult(changed=False, missing=missing)

    layout.mods_dir.mkdir(parents=True, exist_ok=True)
    atomic_write(layout.modlist_txt, expected)
    return _ModlistResult(changed=True, missing=missing)


def _scan_paks(workshop_content: Path, wsid: int) -> list[Path]:
    directory = workshop_content / str(wsid)
    try:
        entries = sorted(directory.iterdir())
    except FileNotFoundError:
        return []
    except NotADirectoryError:
        return []
    except OSError as e:
        raise AxeError("filesystem", f"reading {directory}: {e}") from e
    return [p for p in entries if p.name.lower().endswith(".pak")]


# Backward-compat alias: 0.1 callers used `sync_modlist`.
sync_modlist = run_sync
